aptos-move/framework/aptos-stdlib/sources/smart_vector.move

```diff
@@ -0,0 +1,435 @@
+    public fun singleton<T: store>(e: T): SmartVector<T> {
+        let v = empty();
+        push_back(&mut v, e);
+        v
+    }
+
+    /// Destroy the vector `v`.
+    /// Aborts if `v` is not empty.
+    public fun destroy_empty<T>(v: SmartVector<T>) {
+        assert!(is_empty(&v), error::invalid_argument(EVECTOR_NOT_EMPTY));
+        let SmartVector { inline_inner, table_inner} = v;
+        vector::destroy_empty(inline_inner);
+        vector::destroy_empty(table_inner);
+    }
+
+    /// Acquire an immutable reference to the `i`th element of the vector `v`.
+    /// Aborts if `i` is out of bounds.
+    public fun borrow<T>(v: &SmartVector<T>, i: u64): &T {
+        assert!(i < length(v), error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        let inline_len = vector::length(&v.inline_inner);
+        if (i < inline_len) {
+            vector::borrow(&v.inline_inner, i)
+        } else {
+            big_vector::borrow(vector::borrow(&v.table_inner, 0), i - inline_len)
+        }
+    }
+
+    /// Return a mutable reference to the `i`th element in the vector `v`.
+    /// Aborts if `i` is out of bounds.
+    public fun borrow_mut<T>(v: &mut SmartVector<T>, i: u64): &mut T {
+        assert!(i < length(v), error::invalid_argument(EINDEX_OUT_OF_BOUNDS));
+        let inline_len = vector::length(&v.inline_inner);
+        if (i < inline_len) {
+            vector::borrow_mut(&mut v.inline_inner, i)
+        } else {
+            big_vector::borrow_mut(vector::borrow_mut(&mut v.table_inner, 0), i - inline_len)
+        }
+    }
+
+    /// Empty and destroy the other vector, and push each of the elements in the other vector onto the lhs vector in the
+    /// same order as they occurred in other.
+    /// disclaimer: this function may be costly. Use it at your own discretion.
+    public fun append<T: store>(lhs: &mut SmartVector<T>, other: SmartVector<T>) {
+        let half_len = length(&other) / 2;
+        let i = 0;
+        while (i < half_len) {
+            push_back(lhs, swap_remove(&mut other, i));
+            i = i + 1;
+        };
+        while (length(&other) > 0) {
+            push_back(lhs, pop_back(&mut other));
+        };
+        destroy_empty(other);
+    }
+
+    /// Add element `val` to the end of the vector `v`. It grows the buckets when the current buckets are full.
+    /// This operation will cost more gas when it adds new bucket.
+    public fun push_back<T: store>(v: &mut SmartVector<T>, val: T) {
+        let len = length(v);
+        let inline_len = vector::length(&v.inline_inner);
+        let val_size = size_of_val(&val);
+        if (len == inline_len) {
+            if (val_size * (inline_len + 1) < 150 /* magic number */) {
+                assert!(vector::length(&v.table_inner) == 0, 123);
+                vector::push_back(&mut v.inline_inner, val);
+                return
+            };
+            let estimated_avg_size = max((size_of_val(&v.inline_inner) + val_size) / (inline_len + 1), 1);
```

Comment:
can we generate this estimated_avg_size when creating this smart vector? this value should be the same for every push. doing this once in initialization can save gas

Reply:
we can take the size of the first element as `estimated_avg_size`. But will be very inaccurate for T with variable size. TBH I am not sure which is better.